File: backend/src/shared/scheduling/frequency.py
"""Structured-frequency next-run computation for automations.

Replaces the earlier raw-cron model: a `frequency` dict fully describes a
recurring schedule, and next-run is computed in Python so we can express
intervals cron can't ("every N days / weeks / months"). Shared by the `backend`
API (compute on create/update) and the `server` scheduler (advance after a run).

Weekday convention is **0=Sunday … 6=Saturday** (JS `getDay`), matching the web.

Frequency shapes (the `recurring` detail; `once` stores no frequency and uses an
absolute `run_at`):

    {"kind": "hourly",   "minute": 0}                      # every hour at :MM
    {"kind": "daily",    "time": "09:00"}
    {"kind": "weekdays", "time": "09:00"}                  # Mon–Fri
    {"kind": "weekly",   "weekdays": [1,3,5], "time": "09:00"}
    {"kind": "custom", "unit": "minutely", "interval": 15}
    {"kind": "custom", "unit": "hourly",  "minute": 30}
    {"kind": "custom", "unit": "daily",   "interval": 2, "time": "09:00"}
    {"kind": "custom", "unit": "weekly",  "interval": 2, "weekdays": [1], "time": "09:00"}
    {"kind": "custom", "unit": "monthly", "interval": 1, "monthdays": [1,15], "time": "09:00"}

Sub-daily interval schedules (custom `minutely`, and `hourly` in either form) may
carry an optional daily `window` — they then fire only within a time-of-day span,
realigned to the window's start each day:

    {"kind": "custom", "unit": "minutely", "interval": 15,
     "window": {"start": "09:00", "end": "12:00"}}         # 9:00, 9:15 … 12:00, then next day
    {"kind": "custom", "unit": "hourly", "interval": 2,
     "window": {"start": "09:00", "end": "17:00"}}         # 9:00, 11:00 … 17:00, then next day
"""

from datetime import datetime, timedelta, timezone
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

# Bound the day-by-day search so a mis-specified schedule can't loop forever.
_SEARCH_HORIZON_DAYS = 800

# Floor on "every N minutes" — each fire spawns a full agent session, so guard
# against runaway schedules even if the API is called outside the UI (which
# enforces the same minimum). Mirrors MIN_MINUTELY_INTERVAL in the web lib.
_MIN_MINUTELY_INTERVAL = 5


def _as_utc(value: datetime) -> datetime:
    if value.tzinfo is None:
        return value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc)


def _safe_zone(tz_name: str | None) -> ZoneInfo:
    try:
        return ZoneInfo(tz_name or "UTC")
    except (ZoneInfoNotFoundError, ValueError, KeyError):
        return ZoneInfo("UTC")


def _dow_sun0(d: datetime) -> int:
    """Day-of-week with 0=Sunday (Python's date.weekday() is Mon=0)."""
    return (d.weekday() + 1) % 7


def _parse_hm(time_str: str | None) -> tuple[int, int]:
    try:
        h, m = (time_str or "09:00").split(":")
        return max(0, min(23, int(h))), max(0, min(59, int(m)))
    except (ValueError, AttributeError):
        return 9, 0


def _coerce_int(value: object, default: int, lo: int, hi: int) -> int:
    try:
        return max(lo, min(hi, int(value)))  # type: ignore[arg-type]
    except (ValueError, TypeError):
        return default


def is_valid_frequency(frequency: dict | None) -> bool:
    """Whether `frequency` is a recognized recurring shape."""
    if not isinstance(frequency, dict):
        return False
    kind = frequency.get("kind")
    if kind in ("hourly", "daily", "weekdays"):
        return True
    if kind == "weekly":
        return bool(frequency.get("weekdays"))
    if kind == "custom":
        unit = frequency.get("unit")
        if unit == "minutely":
            return True
        if unit == "hourly":
            return True
        if unit == "daily":
            return True
        if unit == "weekly":
            return bool(frequency.get("weekdays"))
        if unit == "monthly":
            return bool(frequency.get("monthdays"))
    return False


def _parse_window(window: object) -> tuple[int, int] | None:
    """Parse a daily ``{"start","end"}`` span into (start, end) minutes-since-local-
    midnight, or None when absent/degenerate (→ treat as all-day).

    Overnight windows (end <= start, e.g. 22:00–02:00) are not supported and
    collapse to all-day; the UI blocks them so this only guards direct API calls."""
    if not isinstance(window, dict):
        return None
    start = window.get("start")
    end = window.get("end")
    if not isinstance(start, str) or not isinstance(end, str):
        return None
    sh, sm = _parse_hm(start)
    eh, em = _parse_hm(end)
    start_min = sh * 60 + sm
    end_min = eh * 60 + em
    if end_min <= start_min:
        return None
    return start_min, end_min


def _anchored_interval_next(
    step: timedelta, now_local: datetime, base_local: datetime
) -> datetime:
    """Next ``base + k*step`` strictly after now (continuous across days).

    floor((now - base) / step) gives the k with base + k*step <= now, so the next
    slot strictly after now is base + (k+1)*step (holds for any sign of the gap)."""
    n = (now_local - base_local) // step
    return _as_utc(base_local + (n + 1) * step)


def _windowed_interval_next(
    step: timedelta, window: tuple[int, int], now_local: datetime
) -> datetime | None:
    """Next occurrence of an every-`step` schedule confined to a daily window.

    `window` is (start, end) minutes-since-midnight, end inclusive. Occurrences
    each day are start, start+step, … up to and including end; past end the
    schedule resumes at the next day's start (interval realigns per day)."""
    start_min, end_min = window
    # The answer is at most tomorrow's start; bound the search anyway (DST days).
    for i in range(3):
        midnight = (now_local + timedelta(days=i)).replace(
            hour=0, minute=0, second=0, microsecond=0
        )
        start = midnight + timedelta(minutes=start_min)
        end = midnight + timedelta(minutes=end_min)
        if now_local < start:
            candidate = start
        else:
            n = (now_local - start) // step
            candidate = start + (n + 1) * step
        if candidate <= end:
            return _as_utc(candidate)
    return None


def _day_active(frequency: dict, d: datetime, anchor_date: datetime) -> bool:
    """Whether calendar day `d` (a tz-local date) fires under `frequency`."""
    kind = frequency.get("kind")
    dow = _dow_sun0(d)

    if kind == "daily":
        return True
    if kind == "weekdays":
        return dow in (1, 2, 3, 4, 5)  # Mon–Fri
    if kind == "weekly":
        return dow in set(frequency.get("weekdays") or [])

    if kind == "custom":
        unit = frequency.get("unit")
        interval = _coerce_int(frequency.get("interval"), 1, 1, 999)
        if unit == "daily":
            delta = (d.date() - anchor_date.date()).days
            return delta >= 0 and delta % interval == 0
        if unit == "weekly":
            if dow not in set(frequency.get("weekdays") or []):
                return False
            # Phase by the Sunday of each week relative to the anchor's Sunday.
            d_sun = d.date() - timedelta(days=_dow_sun0(d))
            a_sun = anchor_date.date() - timedelta(days=_dow_sun0(anchor_date))
            weeks = (d_sun - a_sun).days // 7
            return weeks >= 0 and weeks % interval == 0
        if unit == "monthly":
            if d.day not in set(frequency.get("monthdays") or []):
                return False
            months = (d.year * 12 + d.month) - (
                anchor_date.year * 12 + anchor_date.month
            )
            return months >= 0 and months % interval == 0
    return False


def compute_next_run(
    frequency: dict | None,
    *,
    tz_name: str | None,
    after: datetime | None = None,
    anchor: datetime | None = None,
) -> datetime | None:
    """Next fire time as an aware UTC instant, or None if the schedule yields none.

    `after` defaults to now; `anchor` (defaults to now) is the phase reference for
    interval schedules ("every N …"). Missed occurrences collapse to the next
    future one (v1 drops misses)."""
    if not is_valid_frequency(frequency):
        return None
    assert frequency is not None
    tz = _safe_zone(tz_name)
    now_local = (_as_utc(after) if after else datetime.now(timezone.utc)).astimezone(tz)
    anchor_local = (
        _as_utc(anchor) if anchor else datetime.now(timezone.utc)
    ).astimezone(tz)

    kind = frequency.get("kind")
    unit = frequency.get("unit") if kind == "custom" else None

    # Sub-daily interval schedules: hourly (preset or custom) and custom minutely,
    # each optionally confined to a daily [start,end] window. These are minute-
    # based, not day-based.
    is_hourly = kind == "hourly" or unit == "hourly"
    is_minutely = unit == "minutely"
    if is_hourly or is_minutely:
        if is_minutely:
            interval = _coerce_int(
                frequency.get("interval"), 15, _MIN_MINUTELY_INTERVAL, 1439
            )
            step = timedelta(minutes=interval)
            base = anchor_local.replace(second=0, microsecond=0)
        else:
            interval = _coerce_int(frequency.get("interval"), 1, 1, 168)
            minute = _coerce_int(frequency.get("minute"), 0, 0, 59)
            step = timedelta(hours=interval)
            base = anchor_local.replace(minute=minute, second=0, microsecond=0)
        # `window` is a custom-only field; presets never carry one.
        window = _parse_window(frequency.get("window")) if kind == "custom" else None
        if window is not None:
            return _windowed_interval_next(step, window, now_local)
        return _anchored_interval_next(step, now_local, base)

    hh, mm = _parse_hm(frequency.get("time"))
    for i in range(_SEARCH_HORIZON_DAYS):
        day = now_local + timedelta(days=i)
        if not _day_active(frequency, day, anchor_local):
            continue
        candidate = day.replace(hour=hh, minute=mm, second=0, microsecond=0)
        if candidate > now_local:
            return _as_utc(candidate)
    return None

File: apps/web/app/dashboard/automation/lib/frequency.ts
import type {
  AutomationFrequency,
  AutomationResponse,
  AutomationTimeWindow,
  CreateAutomationRequest,
} from '@/lib/backend-api';

/** Floor on "every N minutes" — each fire spawns a full agent session, so we
 *  guard against runaway schedules. Mirrors _MIN_MINUTELY_INTERVAL in the
 *  backend (shared/scheduling/frequency.py). */
export const MIN_MINUTELY_INTERVAL = 5;

// Weekday convention: 0=Sunday … 6=Saturday (matches the backend + JS getDay()).
export const DAY_LABELS = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
export const DAY_LABELS_FULL = [
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
];

export type RepeatMode =
  | 'once'
  | 'hourly'
  | 'daily'
  | 'weekdays'
  | 'weekly'
  | 'custom';
export type CustomUnit = 'minutely' | 'hourly' | 'daily' | 'weekly' | 'monthly';

/** A superset draft that holds every mode's fields, so switching Repeat keeps
 *  the user's other entries. Converted to the API's discriminated frequency on
 *  submit. */
export interface ScheduleDraft {
  repeat: RepeatMode;
  runAtLocal: string; // one-time: datetime-local ('yyyy-MM-ddThh:mm', local)
  time: string; // 'HH:MM' (daily/weekdays/weekly + custom daily/weekly/monthly)
  minute: number; // hourly / custom-hourly minute past the hour
  weekdays: number[]; // weekly + custom-weekly (0..6)
  customUnit: CustomUnit;
  interval: number; // custom "every N"
  monthdays: number[]; // custom monthly (1..31)
  // Optional daily time window for sub-daily units (custom minutely/hourly):
  // fire only within [windowStart, windowEnd] local time, realigned to the
  // start each day.
  windowEnabled: boolean;
  windowStart: string; // 'HH:MM'
  windowEnd: string; // 'HH:MM'
  timezone: string; // IANA
}

/** Whether a custom unit is sub-daily (interval-based, window-eligible). */
export function isSubDailyUnit(unit: CustomUnit): boolean {
  return unit === 'minutely' || unit === 'hourly';
}

function pad2(n: number): string {
  return String(n).padStart(2, '0');
}

export function browserTimezone(): string {
  try {
    return Intl.DateTimeFormat().resolvedOptions().timeZone || 'UTC';
  } catch {
    return 'UTC';
  }
}

function toLocalInput(d: Date): string {
  return `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}T${pad2(d.getHours())}:${pad2(d.getMinutes())}`;
}

/** '14:30' → '2:30 PM' for summaries; inputs stay 24h HH:MM. */
export function formatTime12(hhmm: string): string {
  const [h, m] = hhmm.split(':').map(Number);
  if (Number.isNaN(h)) return hhmm;
  const period = h < 12 ? 'AM' : 'PM';
  const h12 = h % 12 === 0 ? 12 : h % 12;
  return `${h12}:${pad2(m || 0)} ${period}`;
}

export function defaultScheduleDraft(): ScheduleDraft {
  return {
    repeat: 'daily',
    runAtLocal: '',
    time: '09:00',
    minute: 0,
    weekdays: [1],
    customUnit: 'daily',
    interval: 1,
    monthdays: [1],
    windowEnabled: false,
    windowStart: '09:00',
    windowEnd: '17:00',
    timezone: browserTimezone(),
  };
}

/** The window slice of a draft, or null when disabled/degenerate (start≥end).
 *  Non-window units never emit one. */
function draftWindow(draft: ScheduleDraft): AutomationTimeWindow | null {
  if (!draft.windowEnabled) return null;
  if (!isWindowValid(draft.windowStart, draft.windowEnd)) return null;
  return { start: draft.windowStart, end: draft.windowEnd };
}

/** 'HH:MM' → minutes since midnight (NaN-safe → -1). */
function hmToMinutes(hhmm: string): number {
  const [h, m] = hhmm.split(':').map(Number);
  if (Number.isNaN(h) || Number.isNaN(m)) return -1;
  return h * 60 + m;
}

/** A window is usable only when the end is strictly after the start
 *  (overnight spans aren't supported). */
export function isWindowValid(start: string, end: string): boolean {
  const s = hmToMinutes(start);
  const e = hmToMinutes(end);
  return s >= 0 && e >= 0 && e > s;
}

/** Reconstruct the draft's window fields from a stored frequency window. */
function windowToDraft(
  window: AutomationTimeWindow | null | undefined,
): Pick<ScheduleDraft, 'windowEnabled' | 'windowStart' | 'windowEnd'> {
  if (!window) return { windowEnabled: false, windowStart: '09:00', windowEnd: '17:00' };
  return { windowEnabled: true, windowStart: window.start, windowEnd: window.end };
}

/** Reconstruct a draft from a stored automation (best-effort for fields the
 *  chosen mode doesn't carry). */
export function automationToDraft(a: AutomationResponse): ScheduleDraft {
  const base = defaultScheduleDraft();
  base.timezone = a.timezone || base.timezone;

  if (a.schedule_kind === 'once') {
    return {
      ...base,
      repeat: 'once',
      runAtLocal: a.next_run_at ? toLocalInput(new Date(a.next_run_at)) : '',
    };
  }

  const f = a.frequency;
  if (!f) return { ...base, repeat: 'daily' };

  if (f.kind === 'hourly') return { ...base, repeat: 'hourly', minute: f.minute };
  if (f.kind === 'daily') return { ...base, repeat: 'daily', time: f.time };
  if (f.kind === 'weekdays') return { ...base, repeat: 'weekdays', time: f.time };
  if (f.kind === 'weekly') {
    return { ...base, repeat: 'weekly', weekdays: f.weekdays, time: f.time };
  }
  // custom
  if (f.unit === 'minutely') {
    return {
      ...base,
      repeat: 'custom',
      customUnit: 'minutely',
      interval: f.interval,
      ...windowToDraft(f.window),
    };
  }
  if (f.unit === 'hourly') {
    return {
      ...base,
      repeat: 'custom',
      customUnit: 'hourly',
      interval: f.interval,
      minute: f.minute,
      ...windowToDraft(f.window),
    };
  }
  if (f.unit === 'daily') {
    return {
      ...base,
      repeat: 'custom',
      customUnit: 'daily',
      interval: f.interval,
      time: f.time,
    };
  }
  if (f.unit === 'weekly') {
    return {
      ...base,
      repeat: 'custom',
      customUnit: 'weekly',
      interval: f.interval,
      weekdays: f.weekdays,
      time: f.time,
    };
  }
  return {
    ...base,
    repeat: 'custom',
    customUnit: 'monthly',
    interval: f.interval,
    monthdays: f.monthdays,
    time: f.time,
  };
}

/** Build the API's discriminated frequency (recurring only). */
export function draftToFrequency(draft: ScheduleDraft): AutomationFrequency | null {
  switch (draft.repeat) {
    case 'hourly':
      return { kind: 'hourly', minute: draft.minute };
    case 'daily':
      return { kind: 'daily', time: draft.time };
    case 'weekdays':
      return { kind: 'weekdays', time: draft.time };
    case 'weekly':
      return { kind: 'weekly', weekdays: draft.weekdays, time: draft.time };
    case 'custom':
      switch (draft.customUnit) {
        case 'minutely':
          return {
            kind: 'custom',
            unit: 'minutely',
            interval: Math.max(MIN_MINUTELY_INTERVAL, draft.interval),
            window: draftWindow(draft),
          };
        case 'hourly':
          return {
            kind: 'custom',
            unit: 'hourly',
            interval: draft.interval,
            minute: draft.minute,
            window: draftWindow(draft),
          };
        case 'daily':
          return {
            kind: 'custom',
            unit: 'daily',
            interval: draft.interval,
            time: draft.time,
          };
        case 'weekly':
          return {
            kind: 'custom',
            unit: 'weekly',
            interval: draft.interval,
            weekdays: draft.weekdays,
            time: draft.time,
          };
        case 'monthly':
          return {
            kind: 'custom',
            unit: 'monthly',
            interval: draft.interval,
            monthdays: draft.monthdays,
            time: draft.time,
          };
      }
    // eslint-disable-next-line no-fallthrough
    default:
      return null;
  }
}

/** The schedule slice of a create/update request. */
export function draftToScheduleApi(
  draft: ScheduleDraft,
): Pick<
  CreateAutomationRequest,
  'schedule_kind' | 'run_at' | 'frequency' | 'timezone'
> {
  if (draft.repeat === 'once') {
    return {
      schedule_kind: 'once',
      run_at: draft.runAtLocal ? new Date(draft.runAtLocal).toISOString() : null,
      frequency: null,
      timezone: draft.timezone || browserTimezone(),
    };
  }
  return {
    schedule_kind: 'recurring',
    run_at: null,
    frequency: draftToFrequency(draft),
    timezone: draft.timezone || browserTimezone(),
  };
}

export function isDraftComplete(draft: ScheduleDraft): boolean {
  if (draft.repeat === 'once') return !!draft.runAtLocal;
  if (draft.repeat === 'weekly') return draft.weekdays.length > 0;
  if (draft.repeat === 'custom') {
    if (draft.customUnit === 'weekly') return draft.weekdays.length > 0;
    if (draft.customUnit === 'monthly') return draft.monthdays.length > 0;
    // A window is optional, but an enabled one must be a valid span.
    if (
      isSubDailyUnit(draft.customUnit) &&
      draft.windowEnabled &&
      !isWindowValid(draft.windowStart, draft.windowEnd)
    ) {
      return false;
    }
  }
  return true;
}

function joinDays(days: number[]): string {
  return [...days]
    .sort((a, b) => a - b)
    .map((d) => DAY_LABELS[d] ?? d)
    .join(', ');
}

function pluralN(n: number, unit: string): string {
  return n === 1 ? unit : `${n} ${unit}s`;
}

/** Append a " · 9:00 AM–12:00 PM" span to a sub-daily summary when windowed. */
function withWindow(
  base: string,
  window: AutomationTimeWindow | null | undefined,
): string {
  if (!window) return base;
  return `${base} · ${formatTime12(window.start)}–${formatTime12(window.end)}`;
}

export function summarizeFrequency(f: AutomationFrequency): string {
  switch (f.kind) {
    case 'hourly':
      return `Hourly at :${pad2(f.minute)}`;
    case 'daily':
      return `Daily at ${formatTime12(f.time)}`;
    case 'weekdays':
      return `Weekdays at ${formatTime12(f.time)}`;
    case 'weekly':
      return `Weekly on ${joinDays(f.weekdays)} at ${formatTime12(f.time)}`;
    case 'custom':
      switch (f.unit) {
        case 'minutely':
          return withWindow(`Every ${pluralN(f.interval, 'minute')}`, f.window);
        case 'hourly':
          // A window supersedes the ":MM" phase (fires are aligned to the
          // window start), so only show ":MM" for the all-day case.
          if (f.window) {
            const every =
              f.interval === 1 ? 'Hourly' : `Every ${f.interval} hours`;
            return withWindow(every, f.window);
          }
          return f.interval === 1
            ? `Hourly at :${pad2(f.minute)}`
            : `Every ${f.interval} hours at :${pad2(f.minute)}`;
        case 'daily':
          return `Every ${pluralN(f.interval, 'day')} at ${formatTime12(f.time)}`;
        case 'weekly':
          return `Every ${pluralN(f.interval, 'week')} on ${joinDays(f.weekdays)} at ${formatTime12(f.time)}`;
        case 'monthly':
          return `Every ${pluralN(f.interval, 'month')} on day ${[...f.monthdays].sort((a, b) => a - b).join(', ')} at ${formatTime12(f.time)}`;
      }
  }
  return 'Custom';
}

/** One-line summary for the list / header. */
export function summarizeSchedule(a: AutomationResponse): string {
  if (a.schedule_kind === 'once') {
    return a.next_run_at
      ? `Once · ${new Date(a.next_run_at).toLocaleString()}`
      : 'Once';
  }
  return a.frequency ? summarizeFrequency(a.frequency) : 'Recurring';
}
